packages/core/src/lib/markdown-it/index.js

```diff
@@ -58,7 +58,38 @@ markdownIt.renderer.rules.fence = (tokens, idx, options, env, slf) => {
   const lang = token.info || '';
   let str = token.content;
   let highlighted = false;
-  let lines;
+  let lines = str.split('\n');
+
+  const startFromOneBased = Math.max(1, parseInt(getAttributeAndDelete(token, 'start-from'), 10) || 1);
+  const startFromZeroBased = startFromOneBased - 1;
+
+  if (startFromOneBased > 1) {
+    // counter is incremented on each span, so we need to subtract 1
+    token.attrJoin('style', `counter-reset: line ${startFromZeroBased};`);
+  }
+
+  const highlightLinesInput = getAttributeAndDelete(token, 'highlight-lines');
+  let highlightRules = [];
+  if (highlightLinesInput) {
+    const highlightLines = highlightLinesInput.split(',');
+    highlightRules = highlightLines.map(HighlightRule.parseRule).filter(rule => rule);
+  }
+  highlightRules.forEach((rule) => {
+    // Note: authors provide line numbers based on the 'start-from' attribute if it exists,
+    //       so we need to shift line numbers back down to start at 0
+    rule.offsetLines(-startFromZeroBased);
```

Comment:
Oh yeah, I hadn't thought of this approach. I can take it one step further so that the result of `parseRuleComponent` is always a properly defined rule component wrt the code block (i.e. already accounting for line offset, figuring out the actual bounds wrt the line, converting to word slices wrt the line, etc), or null if it's not valid.

So, no need to have intermediate properties such as `isWordSlice`, `linePart`, anymore.